### PR TITLE
feat: useStateHistory

### PR DESCRIPTION
**Pull request by xobotyi**
# Description

Introducing hook that stores defined amount of previous state values and provides handles to travel through them

## Type of change

- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [x] New feature _(non-breaking change which adds functionality)_
- [ ] **Breaking change** _(fix or feature that would cause existing functionality to not work as before)_

# Checklist
- [x] Read the [Contributing Guide](https://github.com/streamich/react-use/blob/master/CONTRIBUTING.md)
- [x] Perform a code self-review
- [x] Comment the code, particularly in hard-to-understand areas
- [x] Add documentation
- [x] Add hook's story at Storybook
- [x] Cover changes with tests
- [x] Ensure the test suite passes (`yarn test`)
- [x] Provide 100% tests coverage
- [x] Make sure code lints (`yarn lint`). Fix it with `yarn lint:fix` in case of failure.
- [x] Make sure types are fine (`yarn lint:types`).

### CODE REVIEW

**Comment by streamich** [comment]
What do you think about calling it `useStateWithHistory`?

**Comment by xobotyi** [comment]
Current name is draft.
About your proposal - it would better coney meaning, but for my ears it somewhy sounds not very well.

**Comment by xobotyi** [comment]
I've been thinking about `useHistoryfulState` but not sure about it.

**Comment by xobotyi** [comment]
Finalized and mergable now

**Comment by streamich** [comment]
Reviewing...

**Comment by xobotyi** [comment]
@streamich there been some files got in PR from rebase, deleted them now.

File: docs/useStateWithHistory.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# `useStateHistory`
```
Comment (streamich) [inline docs/useStateWithHistory.md]:
```suggestion
# `useStateWithHistory`
```

Comment (xobotyi) [reply docs/useStateWithHistory.md]: `stateList` is a kinda finite-state machine and works on predefined states list.
While `useStateWithHistory` stores `n` previous states so it is more like fixed size state logger but with travel and modification handles.

I dont think it is possible to combine them, they're about different purposes.

**Pull request merged**